server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -588,6 +622,223 @@ private Scheduler.Cancellable fetchClusterFormationInfo(
+                    node,
+                    node.getVersion(),
+                    minSupportedVersion
+                );
+            } else if (isCancelled.get()) {
+                logger.trace("The remote coordination diagnostics task has been cancelled, so not opening a remote connection");
+            } else {
+                /*
+                 * Since this block is not synchronized it is possible that this task is cancelled between the check above and when the
+                 * code below is run, but this is harmless and not worth the additional synchronization in the normal case. In that case
+                 * the connection will just be closed and the transport request will not be made.
+                 */
+                transportService.connectToNode(
+                    // Note: This connection must be explicitly closed in the connectionListener
+                    node,
+                    ConnectionProfile.buildDefaultConnectionProfile(clusterService.getSettings()),
+                    connectionListener
+                );
+            }
+        }, new TimeValue(10, TimeUnit.SECONDS), ThreadPool.Names.SAME);
+    }
+
+    private void cancelPollingRemoteStableMasterHealthIndicatorService() {
+        synchronized (remoteDiagnosticsMutex) {
+            if (remoteStableMasterHealthIndicatorTask != null && remoteCoordinationDiagnosticsCancelled.getAndSet(true) == false) {
+                remoteStableMasterHealthIndicatorTask.get().cancel();
+                remoteCoordinationDiagnosisResult = new AtomicReference<>();
+            }
+        }
+    }
+
+    @Override
+    public void onFoundPeersUpdated() {
+        /*
+         * If we are on a non-master-eligible node, and the list of peers in PeerFinder is non-empty, that implies that there is
+         * currently no master node elected.
+         * This begins polling a random master-eligible node for its result from this service. However there's a 10-second delay before it
+         * starts, so in the normal situation where during a master transition it flips from master1 -> null -> master2, it the
+         * polling tasks will be canceled before any requests are actually made.
+         * Note that this method can be called from multiple threads.
+         */
+        if (clusterService.localNode().isMasterNode() == false) {
+            /*
+             * Note that PeerFinder (the source of master eligible nodes) could be updating the master eligible nodes on a different
+             * thread, so making a copy here so that it doesn't change for the short duration of this method.
+             */
+            List<DiscoveryNode> masterEligibleNodes = new ArrayList<>(getMasterEligibleNodes());
+            if (masterEligibleNodes.isEmpty()) {
+                cancelPollingRemoteStableMasterHealthIndicatorService();
+            } else {
+                beginPollingRemoteStableMasterHealthIndicatorService(masterEligibleNodes);
+            }
+        }
```

Comment:
I'm wondering. Are these the correct start/stop triggers for this polling? (even further, for the `cancel...` trigger - is the last call for `onFoundPeersUpdated` always coinciding with an empty list of `masterEligibleNodes` in order to satisfy the cancel condition `if (masterEligibleNodes.isEmpty())` ?)

Would it be simpler to start the polling of **one** random master eligible node (as indicated [here](https://user-images.githubusercontent.com/453603/177401768-adda0bc7-112b-4b39-943f-202cfc4d142e.png)) from the cluster state change event notification that illustrates we lost the master node ? And stop the polling when we've got a master? 
(similar to how we do it for the cluster formation fetch/cancel but in the case when the local node is _not_ master eligible)

The `onFoundPeersUpdated` could be used to keep track of the discovered peers in a `CopyOnWriteArrayList` and the polling track could pick a random master eligible node from this list to poll.
(just one would be enough I think - based on the reference diagram - as long as one is reachable, and when it is not, we can source pick another one )

Reply:
The problem is that non-master-eligible nodes do not have _any_ peers in PeerFinder at the time that `clusterChanged` is called with a notification that the master is null. PeerFinder's peers list is populated at some point after that in another thread. I had originally done what you say here, but realized in an integration test (in an upcoming PR) that I would _never_ get any remote diagnostics back because the list of master eligible nodes was _always_ empty, so the polling would stop immediately.

Reply:
Wouldn't the polling retry 10 seconds later? 
ie. if there's nothing in the list of discovered peers we should just retry in 10 seconds and pick a random node from the list of peers at that point ... if any available